### ISSUE TITLE
Gradient checkpoint recomputation updates BatchNorm running statistics twice

### ISSUE THREAD

**Issue by Jpl-hub** [Tencent/YOLO-Master#262]
## Summary

While checking training-state equivalence for activation checkpointing during LoRA memory experiments, I found that `BaseModel._apply_checkpointing()` can update training-mode BatchNorm running statistics again during backward recomputation. The forward values and gradients can still match eager execution, so comparing only losses or gradients misses the state change.

Inspected revision: `2d0a22f7258acfa0f9cfa20db8b2208b10818d60`.

## Minimal reproduction

Run from the repository with PyTorch installed. No pretrained weights, dataset, or GPU is needed.

```python
from copy import deepcopy

import torch
import torch.utils.checkpoint

from ultralytics.nn.modules import Conv
from ultralytics.nn.tasks import BaseModel

torch.manual_seed(0)
torch.set_num_threads(1)
eager = Conv(3, 8, 3).train()
checked = deepcopy(eager)
base = BaseModel().train()

for _ in range(3):
    x = torch.randn(2, 3, 8, 8)
    a = eager(x.clone().requires_grad_())
    b = base._apply_checkpointing(checked, x.clone().requires_grad_())
    torch.testing.assert_close(a, b)
    a.square().mean().backward()
    b.square().mean().backward()

print("BN batches:", eager.bn.num_batches_tracked.item(), checked.bn.num_batches_tracked.item())
print("mean difference:", (eager.bn.running_mean - checked.bn.running_mean).abs().max().item())
print("variance difference:", (eager.bn.running_var - checked.bn.running_var).abs().max().item())
```

Observed output:

```text
BN batches: 3 6
mean difference: 0.012171098962426186
variance difference: 0.162461519241333
```

Expected: both counters are 3 and the running buffers match. Backward recomputation should not add a second persistent update for a training batch.

## Supported LoRA path

I also checked the unchanged upstream dense detector and PEFT injection API:

```python
from types import SimpleNamespace

from ultralytics.nn.tasks import DetectionModel
from ultralytics.utils import DEFAULT_CFG_DICT
from ultralytics.utils.lora.api import apply_lora
from ultralytics.utils.lora.config import LoRAConfig

model = DetectionModel("yolo11n.yaml", nc=2, verbose=False)
model.args = SimpleNamespace(**DEFAULT_CFG_DICT)
model = apply_lora(model, LoRAConfig(
    r=4, alpha=8, dropout=0, backend="peft",
    gradient_checkpointing=True, freeze_bn=False,
))
```

Using two identical model copies, I disabled checkpointing on one and left the API-enabled checkpointing active on the other. Three CPU forward/backward passes used synthetic 2x3x64x64 images and detection labels with the model's detection loss, without optimizer steps:

- 74 BN counters differed after each backward; an affected layer counted 1/2, 2/4, then 3/6 for eager/checkpointed execution.
- Losses and LoRA gradients matched within `rtol=1e-5, atol=1e-6`; 54 LoRA gradient tensors were nonzero on each pass.
- An explicit eval-mode BN control, applied after `model.train()`, had no buffer differences across those batches. This control does not test the trainer's `freeze_bn` lifecycle.

This is a state-correctness test, not an accuracy benchmark. I have not measured an mAP change caused by this issue.

## Scope and possible fix

The relevant implementation is [BaseModel._apply_checkpointing](https://github.com/Tencent/YOLO-Master/blob/2d0a22f7258acfa0f9cfa20db8b2208b10818d60/ultralytics/nn/tasks.py#L220), which invokes non-reentrant checkpointing without protecting BN running buffers during recomputation.

The successful dense PEFT path can enable it. The [core-MoE guard](https://github.com/Tencent/YOLO-Master/blob/2d0a22f7258acfa0f9cfa20db8b2208b10818d60/ultralytics/utils/lora/api.py#L1328) skips checkpoint activation for core-MoE models, so I am not claiming every YOLO-Master model or LoRA backend is affected. Frozen/eval BN is outside this reproduction's affected case.

A possible fix is to isolate adapting BN running buffers during recomputation while preserving normal forward updates. Before proposing a patch, I would cover consecutive batches, cumulative BN, repeated backward, shared buffers, exception cleanup, and the existing routed-module bypass; `torch.compile` compatibility also needs care with custom checkpoint contexts.

Would a focused fix with regression coverage in the existing checkpoint tests be welcome?

## Environment

- Linux x86_64, Python 3.12.3
- PyTorch 2.8.0+cu128; tests above ran on CPU
- PEFT 0.19.1 for the detector integration check

I checked open and closed issue/PR titles and descriptions and did not find the same BN-state report. The MoE/DDP ready-twice issue is a different failure mechanism.
